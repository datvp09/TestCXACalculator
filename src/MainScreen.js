import React from 'react';
import {View, Text, StatusBar, StyleSheet} from 'react-native';
import {
  numberButtons,
  mathButtons,
  isIpX,
  isiOS,
  initialOutput,
} from './constants';
import Button from './Button';
import {connect} from 'react-redux';

const lastMathIndex = mathButtons.length - 1;

const MainScreen = (props) => {
  const _handleEvent = (value) => {
    if (value == 'clear') {
      return props.clear();
    }
    if (value == '=') {
      const re = /(?:(?:^|[-+_*รท])(?:\s*-?\d+(\.\d+)?(?:[eE][+-]?\d+)?\s*))+$/;
      const isValidMathExpression = re.test(props.output);

      if (isValidMathExpression) {
        props.replaceBy('/');
        props.evaluate();
        props.resetMath(true);
      }
      return;
    }

    // reset after calculate
    if (props.reset) {
      if (isNaN(value)) {
        // is operator
        props.concat(value);
      } else {
        // is number
        props.concatWithout(value);
      }
      props.resetMath(false);
      return;
    }

    let strLastChar = props.output.slice(-1);

    // new expression replace old expression
    if (isNaN(strLastChar) && isNaN(value)) {
      props.replace(value);
    } else if (props.output !== initialOutput || isNaN(value)) {
      props.concat(value);
    } else {
      props.concatWithout(value);
    }
  };

  return (
    <View style={[styles.container, isIpX && {marginBottom: 34}]}>
      <StatusBar barStyle={'light-content'} />
      <View style={styles.placeHolderOutput}>
        <Text style={styles.txtDefault}>
          {props.output.toString().replace(/[/]/, 'รท')}
        </Text>
      </View>
      <View style={styles.contRow}>
        <View style={styles.flexNumbers}>
          {numberButtons.map((row, index) => (
            <View key={index} style={styles.contRow}>
              {row.map((col, index) => (
                <Button
                  key={index}
                  textStyle={{
                    color: col == 'clear' ? 'rgba(46,46,46,0.65)' : 'black',
                  }}
                  onPress={() => _handleEvent(col)}
                  text={col}
                />
              ))}
            </View>
          ))}
        </View>
        <View style={styles.container}>
          {mathButtons.map((row, index) => (
            <Button
              style={[
                styles.button,
                {
                  borderTopWidth: index == 0 ? 0 : 1,
                  borderBottomWidth: index == lastMathIndex ? 0 : 1,
                },
              ]}
              textStyle={styles.operator}
              key={index}
              onPress={() => _handleEvent(row)}
              text={row}
            />
          ))}
        </View>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  flexNumbers: {
    flex: 3,
  },
  container: {
    flex: 1,
  },
  contRow: {
    flex: 1,
    flexDirection: 'row',
  },
  placeHolderOutput: {
    flex: 1,
    backgroundColor: '#433437',
    justifyContent: 'flex-end',
    alignItems: 'flex-end',
    padding: 15,
  },
  txtDefault: {
    color: 'white',
    fontFamily: 'System',
    fontSize: 60,
  },
  button: {
    backgroundColor: '#f2274f',
    borderWidth: 0,
    borderTopColor: 'white',
    borderBottomColor: 'white',
  },
  operator: {color: 'white', fontSize: 40, fontWeight: 'bold'},
});

const mapStateToProps = (state) => ({
  output: state.calReducer.output,
  reset: state.calReducer.reset,
});
const mapDispatchToProps = (dispatch) => {
  const {
    clear,
    evaluate,
    replace,
    replaceBy,
    concat,
    concatWithout,
    resetMath,
  } = require('./actions');
  return {
    clear: () => dispatch(clear()),
    evaluate: () => dispatch(evaluate()),
    replace: (value) => dispatch(replace(value)),
    replaceBy: (value) => dispatch(replaceBy(value)),
    concat: (value) => dispatch(concat(value)),
    concatWithout: (value) => dispatch(concatWithout(value)),
    resetMath: (value) => dispatch(resetMath(value)),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(MainScreen);
