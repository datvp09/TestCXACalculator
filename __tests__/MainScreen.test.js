import 'react-native';
import React from 'react';
import MainScreen from '../src/MainScreen';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import configureMockStore from 'redux-mock-store';

const mockStore = configureMockStore();
const store = mockStore({calReducer: {output: '0'}});

describe('MainScreen', () => {
  describe('Rendering', () => {
    it('should match to snapshot', () => {
      const component = renderer
        .create(
          <Provider store={store}>
            <MainScreen />
          </Provider>,
        )
        .toJSON();
      expect(component).toMatchSnapshot();
    });
  });
});
