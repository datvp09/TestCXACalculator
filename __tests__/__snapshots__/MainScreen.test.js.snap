// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainScreen Rendering should match to snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
      false,
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "backgroundColor": "#433437",
        "flex": 1,
        "justifyContent": "flex-end",
        "padding": 15,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontFamily": "System",
          "fontSize": 60,
        }
      }
    >
      0
    </Text>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 3,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "flex": 1,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderColor": "red",
                  "borderWidth": 1,
                  "flex": 1,
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                    "fontFamily": "System",
                    "fontSize": 40,
                    "fontWeight": "500",
                  },
                  Object {
                    "color": "rgba(46,46,46,0.65)",
                  },
                ]
              }
            >
              clear
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "flex": 1,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderColor": "red",
                  "borderWidth": 1,
                  "flex": 1,
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                    "fontFamily": "System",
                    "fontSize": 40,
                    "fontWeight": "500",
                  },
                  Object {
                    "color": "black",
                  },
                ]
              }
            >
              7
            </Text>
          </View>
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "flex": 1,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderColor": "red",
                  "borderWidth": 1,
                  "flex": 1,
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                    "fontFamily": "System",
                    "fontSize": 40,
                    "fontWeight": "500",
                  },
                  Object {
                    "color": "black",
                  },
                ]
              }
            >
              8
            </Text>
          </View>
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "flex": 1,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderColor": "red",
                  "borderWidth": 1,
                  "flex": 1,
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                    "fontFamily": "System",
                    "fontSize": 40,
                    "fontWeight": "500",
                  },
                  Object {
                    "color": "black",
                  },
                ]
              }
            >
              9
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "flex": 1,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderColor": "red",
                  "borderWidth": 1,
                  "flex": 1,
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                    "fontFamily": "System",
                    "fontSize": 40,
                    "fontWeight": "500",
                  },
                  Object {
                    "color": "black",
                  },
                ]
              }
            >
              4
            </Text>
          </View>
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "flex": 1,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderColor": "red",
                  "borderWidth": 1,
                  "flex": 1,
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                    "fontFamily": "System",
                    "fontSize": 40,
                    "fontWeight": "500",
                  },
                  Object {
                    "color": "black",
                  },
                ]
              }
            >
              5
            </Text>
          </View>
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "flex": 1,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderColor": "red",
                  "borderWidth": 1,
                  "flex": 1,
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                    "fontFamily": "System",
                    "fontSize": 40,
                    "fontWeight": "500",
                  },
                  Object {
                    "color": "black",
                  },
                ]
              }
            >
              6
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "flex": 1,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderColor": "red",
                  "borderWidth": 1,
                  "flex": 1,
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                    "fontFamily": "System",
                    "fontSize": 40,
                    "fontWeight": "500",
                  },
                  Object {
                    "color": "black",
                  },
                ]
              }
            >
              1
            </Text>
          </View>
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "flex": 1,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderColor": "red",
                  "borderWidth": 1,
                  "flex": 1,
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                    "fontFamily": "System",
                    "fontSize": 40,
                    "fontWeight": "500",
                  },
                  Object {
                    "color": "black",
                  },
                ]
              }
            >
              2
            </Text>
          </View>
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "flex": 1,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderColor": "red",
                  "borderWidth": 1,
                  "flex": 1,
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                    "fontFamily": "System",
                    "fontSize": 40,
                    "fontWeight": "500",
                  },
                  Object {
                    "color": "black",
                  },
                ]
              }
            >
              3
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "flex": 1,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderColor": "red",
                "borderWidth": 1,
                "flex": 1,
                "justifyContent": "center",
              },
              Array [
                Object {
                  "backgroundColor": "#f2274f",
                  "borderBottomColor": "white",
                  "borderTopColor": "white",
                  "borderWidth": 0,
                },
                Object {
                  "borderBottomWidth": 1,
                  "borderTopWidth": 0,
                },
              ],
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000",
                  "fontFamily": "System",
                  "fontSize": 40,
                  "fontWeight": "500",
                },
                Object {
                  "color": "white",
                  "fontSize": 40,
                  "fontWeight": "bold",
                },
              ]
            }
          >
            ÷
          </Text>
        </View>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "flex": 1,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderColor": "red",
                "borderWidth": 1,
                "flex": 1,
                "justifyContent": "center",
              },
              Array [
                Object {
                  "backgroundColor": "#f2274f",
                  "borderBottomColor": "white",
                  "borderTopColor": "white",
                  "borderWidth": 0,
                },
                Object {
                  "borderBottomWidth": 1,
                  "borderTopWidth": 1,
                },
              ],
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000",
                  "fontFamily": "System",
                  "fontSize": 40,
                  "fontWeight": "500",
                },
                Object {
                  "color": "white",
                  "fontSize": 40,
                  "fontWeight": "bold",
                },
              ]
            }
          >
            *
          </Text>
        </View>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "flex": 1,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderColor": "red",
                "borderWidth": 1,
                "flex": 1,
                "justifyContent": "center",
              },
              Array [
                Object {
                  "backgroundColor": "#f2274f",
                  "borderBottomColor": "white",
                  "borderTopColor": "white",
                  "borderWidth": 0,
                },
                Object {
                  "borderBottomWidth": 1,
                  "borderTopWidth": 1,
                },
              ],
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000",
                  "fontFamily": "System",
                  "fontSize": 40,
                  "fontWeight": "500",
                },
                Object {
                  "color": "white",
                  "fontSize": 40,
                  "fontWeight": "bold",
                },
              ]
            }
          >
            -
          </Text>
        </View>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "flex": 1,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderColor": "red",
                "borderWidth": 1,
                "flex": 1,
                "justifyContent": "center",
              },
              Array [
                Object {
                  "backgroundColor": "#f2274f",
                  "borderBottomColor": "white",
                  "borderTopColor": "white",
                  "borderWidth": 0,
                },
                Object {
                  "borderBottomWidth": 1,
                  "borderTopWidth": 1,
                },
              ],
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000",
                  "fontFamily": "System",
                  "fontSize": 40,
                  "fontWeight": "500",
                },
                Object {
                  "color": "white",
                  "fontSize": 40,
                  "fontWeight": "bold",
                },
              ]
            }
          >
            +
          </Text>
        </View>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "flex": 1,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderColor": "red",
                "borderWidth": 1,
                "flex": 1,
                "justifyContent": "center",
              },
              Array [
                Object {
                  "backgroundColor": "#f2274f",
                  "borderBottomColor": "white",
                  "borderTopColor": "white",
                  "borderWidth": 0,
                },
                Object {
                  "borderBottomWidth": 0,
                  "borderTopWidth": 1,
                },
              ],
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000",
                  "fontFamily": "System",
                  "fontSize": 40,
                  "fontWeight": "500",
                },
                Object {
                  "color": "white",
                  "fontSize": 40,
                  "fontWeight": "bold",
                },
              ]
            }
          >
            =
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
